groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+}
+
+namespace bsl {
+#ifdef BSL_VARIANT_FULL_IMPLEMENTATION
+template <class R, class VISITOR, class VARIANT>
+R visit(VISITOR&& Visitor, VARIANT&& Variant)
+{
+    return BloombergLP::bslstl::VisitImpl<
+        R,
+        bsl::variant_size<typename bsl::remove_reference_t<VARIANT> >::value>(
+        std::forward<VISITOR>(Visitor), std::forward<VARIANT>(Variant));
+}
+
+template <class VISITOR, class VARIANT>
+typename bsl::invoke_result<
+    VISITOR,
+    typename BloombergLP::bslstl::Variant_CVQualAlt0<VARIANT>::type>::type
+visit(VISITOR&& Visitor, VARIANT&& Variant)
+{
+    typedef typename bsl::invoke_result<
+        VISITOR,
+        typename BloombergLP::bslstl::Variant_CVQualAlt0<VARIANT>::type>::type
+        R;
+    return BloombergLP::bslstl::VisitImpl<
+        R,
+        bsl::variant_size<typename bsl::remove_reference_t<VARIANT> >::value>(
+        std::forward<VISITOR>(Visitor), std::forward<VARIANT>(Variant));
+}
+#else  //BSL_VARIANT_FULL_IMPLEMENTATION
+
+template <class R, class VISITOR, class VARIANT>
+R visit(VISITOR& Visitor, VARIANT& Variant)
+{
+    return BloombergLP::bslstl::VisitImpl<R>(Visitor, Variant);
+}
+
+template <class R, class VISITOR, class VARIANT>
+R visit(VISITOR& Visitor, BloombergLP::bslmf::MovableRef<VARIANT> Variant)
+{
+    VARIANT& lvalue = Variant;
+
+    return BloombergLP::bslstl::MoveVisitImpl<R>(Visitor, lvalue);
+}
+
+template <class VISITOR, class VARIANT>
+typename bsl::invoke_result<
+    VISITOR,
+    typename bsl::variant_alternative<0, VARIANT>::type&>::type
+visit(VISITOR& Visitor, VARIANT& Variant)
+{
+    typedef typename bsl::invoke_result<
+        VISITOR,
+        typename bsl::variant_alternative<0, VARIANT>::type&>::type R;
+    return BloombergLP::bslstl::VisitImpl<R>(Visitor, Variant);
+}
+
+template <class VISITOR, class VARIANT>
+typename bsl::invoke_result<
+    VISITOR,
+    BloombergLP::bslmf::MovableRef<
+        typename bsl::variant_alternative<0, VARIANT>::type> >::type
+visit(VISITOR& Visitor, BloombergLP::bslmf::MovableRef<VARIANT> Variant)
+{
+    VARIANT& lvalue = Variant;
+
+    typedef typename bsl::invoke_result<
+        VISITOR,
+        BloombergLP::bslmf::MovableRef<
+            typename bsl::variant_alternative<0, VARIANT>::type> >::type R;
+    return BloombergLP::bslstl::MoveVisitImpl<R>(Visitor, lvalue);
+}
+
+#endif  // BSL_VARIANT_FULL_IMPLEMENTATION
+
+}  // close namespace bsl
+
+namespace BloombergLP {
+namespace bslstl {
+
+                          // ========================
+                          // class Variant_NoSuchType
+                          // ========================
+
+struct Variant_NoSuchType {
+    // This component-private trivial tag type is used to distinguish between
+    // arguments passed by a user, and an 'enable_if' argument.  It is not
+    // default constructible so the following construction never invokes a
+    // constrained single parameter constructor:
+    //..
+    //   variant<SomeType> o(int, {});
```

Comment:
Changed to `variant<SomeType> o(1, {});`
Open to suggestions too. This is a leftover from bsl::optional where the same typo exists because it was badly adapted from `nullopt_t`. The motivation in the comment is questionable here, but i still think it's good practice to not make it default constructible. 
Btw, shall i also fix the comment in bslstl_optional.h ?

Reply:
Can we perhaps change `SomeType` to a known type that obviously isn't constructible from `int`, such as `const char*`? I think that would help make the comment even more clear.

Reply:
It should be something that is constructible from just an integer so the enable_if doesn't SFINAE away the constructor, and something that is a class that can be constructed from an int and a default constructible thing.

Added this as a comment :
```
    // This component-private trivial tag type is used to distinguish between
    // arguments passed by a user, and an 'enable_if' argument.  It is not
    // default constructible so the following construction never invokes a
    // constrained single parameter constructor:
    //..
    //
    //   struct SomeType
    //   {
    //       SomeType(int, const std::string &s = "default"){}
    //   };
    //
    //   variant<SomeType> o(1, {});      
    //..
```

These are all minor issues. I'll gather all optional stuff and do a pull at the end of out review.

Reply:
right, my bad, I got confused about what it was for. This version of the comment is good.